Fix control flow register mapping in `BasisTranslator`

Cryoris opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This is a manual backport of https://github.com/Qiskit/qiskit/pull/15143 to `stable/1.4` since the file locations and Rust interface was different back then.

### Details and comments

Fixes #13728.

raynelfss committed:
Backport e9f8ffeda

Co-authored-by: Raynel Sanchez <raynelfss@hotmail.com>

Cryoris committed:
Merge branch 'stable/1.4' into backport-basis

Cryoris committed:
Reno

Cryoris committed:
fix clippy

jakelishman commented:
Should this have a test to verify the backport too?

Cryoris commented:
Oh yeah, I do have a test locally but it didn't make the commit 🙂 Incoming!

Cryoris committed:
Add tests

raynelfss reviewed (dismissed):
LGTM!

raynelfss commented on crates/accelerate/src/basis/basis_translator/mod.rs:
You might have missed the mapping that happens here since it was added to the branch after the last time you saw it.

```suggestion
        let unique_qargs: Option<Qargs> = if qubit_set.is_empty() {
            None
        } else {
             // Map the unique qargs with the absolute indices as well
            if let Some(qarg_mapping) = qarg_mapping {
                Some(qubit_set
                    .iter()
                    .map(|x| PhysicalQubit(qarg_mapping[x].0))
                    .collect())
            } else {
                Some(qubit_set.iter().map(|x| PhysicalQubit(x.0)).collect())
            }
        };
```
Without this the translation might fail claiming that a gate wasn't mapped correctly.

raynelfss replied:
You can also add a test case for the nested case similar to the one I send you offline (We'd have to add this on main too).

```py
from qiskit.circuit import QuantumCircuit, QuantumRegister, ClassicalRegister, Parameter
from qiskit.circuit.gate import Gate
from qiskit.providers.fake_provider import GenericBackendV2
from qiskit.transpiler.passes.basis import BasisTranslator
from qiskit.circuit.equivalence_library import StandardEquivalenceLibrary as std_eq_lib

# Create 3 qubit backend
backend = GenericBackendV2(num_qubits=3, control_flow=True)

class TwoQubitOneParamGate(Gate):
    """Mock two qubit one param gate."""

    def __init__(self, theta, name="2q1p"):
        super().__init__(name, 2, [theta])


# Add only this pair as native to the Target.
pair = (0, 2)
alpha = Parameter("alpha")
gate = TwoQubitOneParamGate(alpha)

# Add custom gate to be supported by the Target.
backend.target.add_instruction(gate, {pair: None})  # properties omitted for brevity

qreg = QuantumRegister(3, "data")
creg = ClassicalRegister(1, "res")
circuit = QuantumCircuit(qreg, creg)
circuit.h(0)
circuit.cy(0,2)
circuit.measure(0,0)
with circuit.if_test((creg[0], 0)) as else_:
    with circuit.for_loop(range(2)):
        # circuit.cx(0,2)
        circuit.append(gate, [qreg[0], qreg[2]])
with else_:
    circuit.h(0)
    circuit.cy(0,2)

result = BasisTranslator(std_eq_lib, [], backend.target)(circuit)
```

raynelfss reviewed (commented):
Here's something I caught last minute right after approving this PR.

Cryoris committed:
Fix nested loop and randomized orders

Since the extra instruction attributes stored the
indices in randomized order, the lookup would sometimes
fail randomly. This has already been fixed in the
2.x series.

Co-authored-by: Raynel Sanchez <raynelfss@hotmail.com>
